File: code/src/main_app/ui/GrapherWidget.cpp
#include "main_app/ui/GrapherWidget.hpp"

#include <spdlog/spdlog.h>

#include <QPainter>

#include "main_app/events_manager/EventManager.hpp"

namespace ui::internal {
GrapherWidget::GrapherWidget(QWidget *parent) : QWidget(parent) {
    events_manager::EventManager::getInstance().signalNewFrame.connect([this](const grapher::DrawableSimulation& state) {
        this->onNewFrame(state);
    });
}

void GrapherWidget::paintEvent(__attribute__((unused)) QPaintEvent * event) {
    const double MARGIN = 20.0;
    const double WIDTH = width() - MARGIN * 2;
    const double HEIGHT = height() - MARGIN * 2;
    const double PARTICLE_RADIUS = 10.0;

      QPainter painter(this);
    painter.setRenderHint(QPainter::Antialiasing);

    painter.setBrush(Qt::black);
    for(const auto& particlePosition : frame.particlePositions) {
        const double drawnPositionX = MARGIN + particlePosition.x * WIDTH;
        const double drawnPositionY = MARGIN + particlePosition.y * HEIGHT;
        painter.drawEllipse(QPointF(drawnPositionX, drawnPositionY),  PARTICLE_RADIUS, PARTICLE_RADIUS);
        spdlog::debug("Drawn ({}, {})", drawnPositionX, drawnPositionY);
    }

}

QSize GrapherWidget::sizeHint() const {
    constexpr int MIN_WIDTH = 20;
    constexpr int MIN_HEIGHT = 20;

    if (width() < MIN_WIDTH || height() < MIN_HEIGHT) {
        return {MIN_WIDTH, MIN_HEIGHT};
    }

    return {width(), height()};
}

void GrapherWidget::onNewFrame(const grapher::DrawableSimulation& state) {
    spdlog::debug("New frame!");
    this->frame = state;
    update();
}
}

File: code/constraint_based_simulator/ui/GrapherWidget.py
from PySide6.QtCore import QSize, QRect, Qt, Signal, Slot
from PySide6.QtGui import QPen, QBrush, QPalette, QPainter, QResizeEvent, QColorConstants, QPaintEvent
from PySide6.QtWidgets import QWidget
from typing_extensions import Union

from constraint_based_simulator.common.MainLogger import MAIN_LOGGER
from constraint_based_simulator.events_manager import InitializationSignals
from constraint_based_simulator.grapher.drawables.BarDrawable import BarDrawable
from constraint_based_simulator.grapher.drawables.CircleDrawable import CircleDrawable
from constraint_based_simulator.grapher.drawables.DrawableScene import DrawableScene
from constraint_based_simulator.grapher.drawables.PointDrawable import PointDrawable


class GrapherWidget(QWidget):
    """
    Grapher Qt Widget, view for the simulator (model-view-presenter pattern)
    """
    newFrame = Signal(DrawableScene)

    def __init__(self, parent: Union[QWidget, None] = None) -> None:
        super().__init__(parent)

        self.scene = DrawableScene([])

        self.pen = QPen()
        self.brush = QBrush()
        self.brush.setColor(QColorConstants.Black)  # type: ignore[attr-defined]
        self.brush.setStyle(Qt.BrushStyle.SolidPattern)

        self.setBackgroundRole(QPalette.Base)  # type: ignore[attr-defined]
        self.setAutoFillBackground(True)

        self.newFrame.connect(self.onNewFrame)

    def minimumSizeHint(self) -> QSize:
        return QSize(100, 100)

    def sizeHint(self) -> QSize:
        return QSize(400, 200)

    def resizeEvent(self, event: QResizeEvent) -> None:
        InitializationSignals.grapherParameters.emit(event.size().width(), event.size().height())

    @Slot(DrawableScene)  # type: ignore
    def onNewFrame(self, drawableScene: DrawableScene) -> None:  # pylint: disable=missing-function-docstring
        self.scene = drawableScene
        MAIN_LOGGER.debug(self.scene)
        self.update()

    def paintEvent(self, _: QPaintEvent) -> None:
        # TODO fix the positioning inside the window

        with QPainter(self) as painter:
            painter.setPen(self.pen)
            painter.setBrush(self.brush)
            painter.setRenderHint(QPainter.Antialiasing)  # type: ignore[attr-defined]

            for drawable in self.scene.allDrawables:
                if isinstance(drawable, PointDrawable):
                    radius = 2
                    painter.drawEllipse(drawable.x - radius, drawable.y - radius, radius*2, radius*2)
                    painter.drawText(drawable.x, drawable.y, drawable.identifier)
                elif isinstance(drawable, BarDrawable):
                    painter.drawLine(drawable.xStart, drawable.yStart, drawable.xEnd, drawable.yEnd)
                elif isinstance(drawable, CircleDrawable):
                    radius = drawable.radius
                    painter.setBrush(Qt.NoBrush)  # type: ignore[attr-defined]
                    painter.drawEllipse(drawable.x - radius, drawable.y - radius, radius*2, radius*2)
                    painter.setBrush(self.brush)

            painter.setPen(self.palette().dark().color())
            painter.setBrush(Qt.NoBrush)  # type: ignore[attr-defined]
            painter.drawRect(QRect(0, 0, self.width() - 1, self.height() - 1))
